ligo/src/proposal.mligo

```diff
@@ -4,21 +4,361 @@
+      then
+        if proposal.upvotes > proposal.downvotes
+        then
+          let store = unfreeze_proposer_token (config, true, proposal, proposal_key, store) in
+          let store = unfreeze_voter_token (proposal, proposal_key, store) in
+          let (new_ops, store) = config.decision_lambda (proposal, store) in
+          let cons = fun (l, e : operation list * operation) -> e :: l in
+          let ops = List.fold cons ops new_ops in
+          let store = delete_proposal (start_date, proposal_key, store) in
+          (ops, store, counter)
+        else
+          let store = unfreeze_proposer_token (config, false, proposal, proposal_key, store) in
+          let store = unfreeze_voter_token (proposal, proposal_key, store) in
+          let store = delete_proposal (start_date, proposal_key, store) in
+          (ops, store, counter)
+      else
+        let store = unfreeze_proposer_token (config, false, proposal, proposal_key, store) in
+        let store = unfreeze_voter_token (proposal, proposal_key, store) in
+        let store = delete_proposal (start_date, proposal_key, store) in
+        (ops, store, counter)
+    else (ops, store, counter)
+  else (ops, store, counter)
+
+// Flush all proposals that passed their voting period.
+let flush(n, config, store : nat * config * storage): return =
+  let store = ensure_not_migrated store in
+  let store =
+    if n = 0n
+      then (failwith("BAD_ENTRYPOINT_PARAMETER") : storage)
+      else store
+    in
+
+  let counter : counter =
+    { current = 0n
+    ; total = n
+    } in
+  let flush_one
+      (acc, e: (operation list * storage * counter) * (timestamp * proposal_key)) =
+        let (ops, store, counter) = acc in
+        let (start_date, proposal_key) = e in
+        handle_proposal_is_over (config, start_date, proposal_key, store, ops, counter)
+      in
+  let (ops, store, counter) =
+    Set.fold flush_one store.proposal_key_list_sort_by_date (([] : operation list), store, counter) in
+  (ops, store)
+
+// Removes an accepted and finished proposal by key.
+let drop_proposal (proposal_key, config, store : proposal_key * config * storage): return =
+  let store = ensure_not_migrated store in
+  let store = authorize_admin store in
+
+  let proposal = check_if_proposal_exist (proposal_key, store) in
+  if is_voting_period_over(proposal, store)
+  then
+    if do_total_vote_meet_quorum_threshold(proposal, store)
+    then
+      if proposal.upvotes > proposal.downvotes
+      then
+        let store = unfreeze_proposer_token (config, true, proposal, proposal_key, store) in
+        let store = unfreeze_voter_token (proposal, proposal_key, store) in
+        let store = delete_proposal (proposal.start_date, proposal_key, store) in
+        (([] : operation list), store)
+      else (failwith("FAIL_DROP_PROPOSAL_NOT_ACCEPTED") : return)
+    else (failwith("FAIL_DROP_PROPOSAL_NOT_ACCEPTED") : return)
+  else (failwith("FAIL_DROP_PROPOSAL_NOT_OVER") : return)
```

Comment:
```suggestion
  if   is_voting_period_over(proposal, store)
    && do_total_vote_meet_quorum_threshold(proposal, store)
    && proposal.upvotes > proposal.downvotes
  then
    let store = unfreeze_proposer_token (config, true, proposal, proposal_key, store) in
    let store = unfreeze_voter_token (proposal, proposal_key, store) in
    let store = delete_proposal (proposal.start_date, proposal_key, store) in
    (([] : operation list), store)
  else (failwith("FAIL_DROP_PROPOSAL_NOT_OVER") : return)
```

Reply:
I was also thinking about doing a similar optimization, but to be on par with Lorentz version we will also have to update `flush` in Lorentz. Though we want to do this optimization there for some time already, let's do it now.

Reply:
I think it's fine to make some optimization on the fly if we notice them, in some cases these can be easier to spot in LIGO (not necessarily this case) and vice-versa.

There's also the fact that parts of the contracts will need to be different and so their optimizations too, so in general I don't think there's a problem with them (although we want to remember this if we need/ to optimize the Lorentz version too).